vllm/distributed/device_communicators/shm_broadcast.py

```diff
@@ -0,0 +1,119 @@
+import time
+import multiprocessing
+from multiprocessing import shared_memory, resource_tracker
+from multiprocessing import Array, Value
+import contextlib
+from contextlib import contextmanager
+from torch.distributed import ProcessGroup
+import torch
+import torch.distributed as dist
+import pickle
+
+
+class ShmRingBuffer:
+
+    def __init__(self, pg: ProcessGroup, max_chunk_bytes, max_chunks):
+        self.rank = dist.get_rank(pg)
+        self.world_size = dist.get_world_size(pg)
+        global_ranks = dist.get_process_group_ranks(pg)
+        self.is_writer = self.rank == 0
+        self.is_reader = not self.is_writer
+        self.current_idx = 0
+        self.max_chunk_bytes = max_chunk_bytes
+        self.max_chunks = max_chunks
+        total_bytes = (self.max_chunk_bytes +
+                       self.world_size) * self.max_chunks
+        self.buffer_offset = 0
+        self.metadata_offset = self.max_chunk_bytes * self.max_chunks
+
+        if self.is_writer:
+            self.shared_memory = shared_memory.SharedMemory(create=True,
+                                                            size=total_bytes)
+            # initialize the buffer to 0
+            for i in range(total_bytes):
+                self.shared_memory.buf[i] = 0
+            dist.broadcast_object_list([self.shared_memory.name],
+                                       src=global_ranks[0])
+        else:
+            recv = [None]
+            dist.broadcast_object_list(recv, src=global_ranks[0])
+            name = recv[0]
+            self.shared_memory = shared_memory.SharedMemory(name=name)
+            resource_tracker.unregister(self.shared_memory._name,
+                                        "shared_memory") # noqa
+
+    @property
+    def buffer(self):
+        start = self.buffer_offset + self.current_idx * self.max_chunk_bytes
+        end = start + self.max_chunk_bytes
+        return memoryview(self.shared_memory.buf[start:end])
+
+    @property
+    def metadata(self):
+        start = self.metadata_offset + self.current_idx * self.world_size
+        end = start + self.world_size
+        return memoryview(self.shared_memory.buf[start:end])
+
+    @contextmanager
+    def acquire_write(self):
+        assert self.is_writer, "Only writers can acquire write"
+        while True:
```

Comment:
suggest adding `time.sleep` to busy loops to let OS switch if needed

Reply:
fixed in [a105688](https://github.com/vllm-project/vllm/pull/5399/commits/a105688f1b44b369238d81204bc201abd004a904)